sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/AutoCheckpointer.java

```diff
@@ -42,15 +43,22 @@ public void close(ChangeFeedObserverContext context, ChangeFeedObserverCloseReas
     }
 
     @Override
-    public void processChanges(ChangeFeedObserverContext context, List<CosmosItemProperties> docs) {
-        this.observer.processChanges(context, docs);
+    public Mono<Void> processChanges(ChangeFeedObserverContext context, List<CosmosItemProperties> docs) {
+        return this.observer.processChanges(context, docs)
+            .then(this.afterProcessChanges(context));
+    }
+
+    private Mono<Void> afterProcessChanges(ChangeFeedObserverContext context) {
```

Comment:
now that this method is changed to async, there are some questions around thread-safety/concurrency.

class fields (`lastCheckpointTime`, `lastCheckpointTime`, etc) here are accessed on multiple threads (how are we ensuring thread-safety) on these?

Reply:
Good catch. Normally there's one thread processing per each partition key but there could be multiple feed responses and the process times for each can vary, with an earlier feed response potentially finishing later than a later one which could result in an earlier token being saved in the corresponding lease. I believe restricting the feeds to one at a time in oder to preserve the ordering should take care of it. I'll confirm this with the .NET ppl to see if this is a reasonable solution.

Reply:
Change is in.